Sources/NIO/Codec.swift

```diff
@@ -980,52 +983,69 @@ public class NIOSingleStepByteToMessageProcessor<Decoder: NIOSingleStepByteToMes
     }
 
     private func append(_ buffer: ByteBuffer) {
-        if self.buffer == nil {
-            self.buffer = buffer
+        if self._buffer == nil || self._buffer!.readableBytes == 0 {
+            self._buffer = buffer
         } else {
             var buffer = buffer
-            self.buffer!.writeBuffer(&buffer)
+            self._buffer!.writeBuffer(&buffer)
+        }
+    }
+
+    private func withNonCoWBuffer(_ body: (inout ByteBuffer) throws -> Decoder.InboundOut?) throws -> Decoder.InboundOut? {
+        guard var buffer = self._buffer else {
+            return nil
         }
+
+        if buffer.readableBytes == 0 {
+            return nil
+        }
+
+        self._buffer = nil  // To avoid CoW
+        let result = try body(&buffer)
+        self._buffer = buffer
+        return result
     }
 
     private func decodeLoop(decodeMode: DecodeMode, seenEOF: Bool = false, _ messageReceiver: (Decoder.InboundOut) throws -> Void) throws {
         // we want to call decodeLast once with an empty buffer if we have nothing
-        if decodeMode == .last && (self.buffer == nil || self.buffer!.readableBytes == 0) {
-            var emptyBuffer = self.buffer == nil ? ByteBufferAllocator().buffer(capacity: 0) : self.buffer!
+        if decodeMode == .last && (self._buffer == nil || self._buffer!.readableBytes == 0) {
+            var emptyBuffer = self._buffer == nil ? ByteBufferAllocator().buffer(capacity: 0) : self._buffer!
             if let message = try self.decoder.decodeLast(buffer: &emptyBuffer, seenEOF: seenEOF) {
                 try messageReceiver(message)
             }
             return
         }
 
         // buffer can only be nil if we're called from finishProcessing which is handled above
-        assert(self.buffer != nil)
-
-        var decodingState: DecodingState = .continue
-        while self.buffer!.readableBytes > 0 && decodingState == .continue {
-            let messageOpt: Decoder.InboundOut?
-            if decodeMode == .normal {
-                messageOpt = try self.decoder.decode(buffer: &self.buffer!)
-                if messageOpt == nil, let maximumBufferSize = self.maximumBufferSize, self.buffer!.readableBytes > maximumBufferSize {
-                    throw ByteToMessageDecoderError.PayloadTooLargeError()
+        assert(self._buffer != nil)
+
+        var messageOpt: Decoder.InboundOut?
+        repeat {
+            messageOpt = try self.withNonCoWBuffer() { buffer in
+                defer {
+                    if buffer.readableBytes > 0 {
+                        if self.decoder.shouldReclaimBytes(buffer: buffer) {
+                            buffer.discardReadBytes()
+                        }
+                    }
+                }
+                if decodeMode == .normal {
+                    return try self.decoder.decode(buffer: &buffer)
+                } else {
+                    return try self.decoder.decodeLast(buffer: &buffer, seenEOF: seenEOF)
                 }
```

Comment:
If you split the body of the closure passed to `withNonCoWBuffer` into a separate function, we can rewrite this loop to use `while let` instead of `repeat...while`, which will remove yet another check on the nullity of `messageOpt`.